### ISSUE TITLE
Gazeteer support for StringMatchingRecommender

### ISSUE THREAD

**Issue by reckart** [inception-project/inception#914]
**Is your feature request related to a problem? Please describe.**
Many named entities can be found simply by looking them up in a list. It can be quite tedious to have to mark them in actual text before the StringMatchingRecommender can suggest them.

**Describe the solution you'd like**
It would be nice if StringMatchingRecommenders could be primed using some kind of gazeteer file.

**Describe alternatives you've considered**
One could also consider implementing a new type of GazeteerRecommender.

**Comment by Rentier** [comment]
I would prefer to implement a new type of recommender.

**Comment by reckart** [comment]
Why?

**Comment by Rentier** [comment]
Because we would mix two kinds of recommenders into one which also need different configurations. We can factor out the trie stuff as a common base.

**Comment by reckart** [comment]
I'm not sure. The recommender still a string matching recommender - the only different is that there is data to pre-train on. This is similar to providing pre-trained documents to e.g. some sequence recommender.

**Comment by Rentier** [comment]
For me that are two different use cases, the first one is to project labels from your already existing annotations and the other to annotate stuff based on a predefined dictionary. What do you do if you just want to annotate from a dictionary and not project annotations?

**Comment by reckart** [comment]
> What do you do if you just want to annotate from a dictionary and not project annotations?

Valid point. I wonder though if / why I would want to do that. If I had that case, I would probably add a setting "trainable" to the `StringMatchingRecommender` where the user can choose to only rely on the Gazeteers or also learns dynamically.

From my perspective, the use-case is a bit different though. Right now, the only way that INCEpTION allows the use of pre-trained data is by importing pre-annotated documents, then blocking these documents for annotators and then making sure that the recommender still trains on the blocked document state. But for word-lists, it is quite tedious and annoying to first have to convert them into pseudo-annotated documents. It is much more convenient if I can import them as word lists and use them. So the idea was to have an alternative way of pre-training the StringMatchingRecommender - not via pre-annotated documents but rather via convenient word lists.

If we pre-train the `StringMatchingRecommender` with the word lists, then the score provided by the recommender will consider the word list as one observation - thus if we have `Mercedes` as `PER` in the word list and then the user annotates `Mercedes` as `ORG` in the text twice, we get a score of `0.66` for `ORG` and `0.33` for `PER`. If we use a completely different recommender, then Active Learning cannot even properly compare the scores across recommenders because we cannot assume that the score scales are the same for all recommenders. AL currently only considers differences in scores produces by the same recommender - so it could work okish with a pre-trained StringMatchingRecommender, but it would not work well if we use two different types of recommenders.

**Comment by reckart** [comment]
That said, it might be worth considering the addition of a `prior` column to the gazeteer at some point.

**Comment by Rentier** [comment]
I think that you address problems that we have to solve for all recommenders instead of pointwise:
- We want to aloww pretraining for all recommenders 
- We want all recommenders to interact nicely when doing active learning on the same feature

I have to say that I really do not like the mix of recommenders here, I would prefer to seperating their concerns.

**Comment by reckart** [comment]
We do already support pre-training for all recommenders by:

* uploading pre-annotated documents
* blocking them for user annotation
* configuring recommenders to train on them

But it is not a convenient pre-training mode for dictionary-based recommenders such as the `StringMatchingRecommender` where it is much easier to upload a dictionary to start with.

Do you have an idea how we could make confidence scores compatible between recommenders? The scores depend on the underlying algorithm and the confidence scores provided by an SVM, a Neural Network, a Naive Bayes, etc. tend to behave differently.

**Comment by Rentier** [comment]
For some models, I would like to use a pretrained model, not pretrain it on my own data, that is what I meant with my first point. I thought that is closer to your use case. I agree it is more practical to use an external source for dictionary based recommenders.

For the probabillities: I do not have a good solution for  that right now, but I do not think that that is a reason for mixing two recommenders. You can do it this way if you want, but I find that it bloats the string recommender more without too much of an advantage.

**Comment by reckart** [comment]
Let's assume you wanted to use a pre-trained model with the OpenNLP NER Recommender - how would you do that?

**Comment by Rentier** [comment]
The first basic version be: point to a serialized form of the pretrained model and load it, when training, then train on the new training data. Normally, one would need to train on all data, which might be infeasible for large pretraining data. Hopefully, the new data is enough to adjust it to our domain.

**Comment by reckart** [comment]
Would you create a new `PreTrainedOpenNlpNerRecommender` or extend the existing `OpenNlpNerRecommender`?

**Comment by Rentier** [comment]
I would add a pretraining hook to the recommender engine and then roll it out wherever it is possible.

**Comment by reckart** [comment]
If you train a dictionary-based recommender, what you get is a dictionary. If you save that do disk, you basically get a gazeteer.

**Comment by reckart** [comment]
Thus, loading a dictionary into a dictionary-based recommender amounts to loading a pre-trained model, right?

**Comment by Rentier** [comment]
Yes, but my issue is then that they are different in that a gazeteer is not trained on new annotations, but a string based recommender is.

**Comment by reckart** [comment]
Let's consider the following scenario:

* I have a project containing some texts and I annotate them - I am using the `OpenNlpNerRecommender` to help me with that
* now let's assume we had an option to export recommender models so we could share them with people as "pre-trained models"
* so I export the model and send it to you
* you import it as a pre-trained into a `OpenNlpNerRecommender` which you are using on your own computer to help you in your own annotation

If you agree to the story above, then simply replace `OpenNlpNerRecommender` with `StringMatchingRecommender` and consider  `gazeteer` to the file format of the `pre-trained model` (as compared e.g to OpenNLP NER model for `OpenNlpNerRecommender`). Makes sense?

**Comment by reckart** [comment]
I'll give it a day or two to thinking more about it. 

But I have to admit that any doubts about the current solution which may have caused me to mention the `GazeteerRecommender` as an alternative have rather been resolved than strengthened.
